### DISCUSSION TITLE
Issues with GrayDiffuseRadiation Convergence

### DISCUSSION THREAD

**Discussion by mrplants** [idaholab/moose#23590] [Q&A General] [answered]
I am attempting to use GrayDiffuseRadiation to model the transfer of heat from an outer cylinder (stator) and inner cylinder (rotor).  Here is an illustration of my mesh.  (The orange section is the vacuum):

https://user-images.githubusercontent.com/3487464/222600814-2257d69e-8a87-47fe-a522-bf311c8853b5.mov

Here is my input file:

```
[Mesh]
  type = MeshGeneratorMesh
  [file_mesh]
    type = FileMeshGenerator
    file = cylinder_within_cylinder.msh
  []
  [rotor_outer_surface_sideset]
    type = SideSetsBetweenSubdomainsGenerator
    input = file_mesh
    primary_block = 'rotor_volume'
    paired_block = 'vacuum_volume'
    new_boundary = 'rotor_outer_surface_sideset'
  []
  [stator_inner_surface_sideset]
    type = SideSetsBetweenSubdomainsGenerator
    input = rotor_outer_surface_sideset
    primary_block = 'stator_volume'
    paired_block = 'vacuum_volume'
    new_boundary = 'stator_inner_surface_sideset'
  []
  []

[Variables]
  [Temperature]
  []
[]

[ICs]
  [rotor_IC]
    type = ConstantIC
    variable = Temperature
    block = 'rotor_volume'
    value = 80
  []
  [vacuum_IC]
    type = ConstantIC
    variable = Temperature 
    block = 'vacuum_volume'
    value = 80
  []
  [stator_IC]
    type = ConstantIC
    variable = Temperature 
    block = 'stator_volume'
    value = 300
  []
[]

[Kernels]
  [heat_conduction]
    type = HeatConduction
    variable = Temperature
  []
  [heat_conduction_time_derivative]
    type = HeatConductionTimeDerivative
    variable = Temperature
  []
[]

[GrayDiffuseRadiation]
  [vacuum_radiation]
    boundary = 'stator_inner_surface_sideset rotor_outer_surface_sideset'
    emissivity = '0.03 0.03'
    n_patches = '2 2'
    temperature = Temperature
    ray_tracing_face_type = 'GAUSS'
  []
[]

[Materials]
  [thermal_rotor]
    type = HeatConductionMaterial
    thermal_conductivity = 45.0
    specific_heat = 0.5
    block = 'rotor_volume'
  []
  [thermal_vacuum]
    type = HeatConductionMaterial
    thermal_conductivity = 0
    specific_heat = 0
    block = 'vacuum_volume'
  []
  [thermal_stator]
    type = HeatConductionMaterial
    thermal_conductivity = 25.0
    specific_heat = 1
    block = 'stator_volume'
  []
  [density]
    type = GenericConstantMaterial
    prop_names = 'density'
    prop_values = 8000.0
  []
[]

[Executioner]
  type = Transient
  end_time = 5
  dt = 1
[]

[Outputs]
  exodus = true
[]
```

When I run this, it does not converge.  I have searched high and low and cannot fathom what could be wrong in such a simple problem.  What is going on here?

**Comment by mrplants** [comment]
GitHub won't allow me to upload my mesh (.msh) or input (.i) files.  Any recommendations?

**Reply by GiudGiud** [reply]
You should be able to upload them to a repository and share the link to it with us here

**Reply by mrplants** [reply]
Here you go:
[repo](https://github.com/inertial-energy/vacuum_thermal)

**Comment by GiudGiud** [comment, marked as answer]
@loganharbour @snschune 
My guess is that 2x2 patches is not enough for 2 concentric cylinders but you ll know more

**Reply by mrplants** [reply]
Just tried increasing to 20x20 and it still does not converge.

**Reply by snschune** [reply]
The temperature variable was defined in the vacuum region with `cp=0` and `k=0` and that led to numerical difficulties. 
This one should work:

```
[Mesh]
  type = MeshGeneratorMesh
  [file_mesh]
    type = FileMeshGenerator
    file = cylinder_within_cylinder.msh
  []
  [rotor_outer_surface_sideset]
    type = SideSetsBetweenSubdomainsGenerator
    input = file_mesh
    primary_block = 'rotor_volume'
    paired_block = 'vacuum_volume'
    new_boundary = 'rotor_outer_surface_sideset'
  []
  [stator_inner_surface_sideset]
    type = SideSetsBetweenSubdomainsGenerator
    input = rotor_outer_surface_sideset
    primary_block = 'stator_volume'
    paired_block = 'vacuum_volume'
    new_boundary = 'stator_inner_surface_sideset'
  [../]
  []

[Variables]
  [Temperature]
    block = 'rotor_volume stator_volume'
  []
[]

[Problem]
  kernel_coverage_check = false
  material_coverage_check = false
[]

[ICs]
  [rotor_IC]
    type = ConstantIC
    variable = Temperature
    block = 'rotor_volume'
    value = 80
  []
  [stator_IC]
    type = ConstantIC
    variable = Temperature
    block = 'stator_volume'
    value = 300
  []
[]

[Kernels]
  [heat_conduction]
    type = HeatConduction
    variable = Temperature
  []
  [heat_conduction_time_derivative]
    type = HeatConductionTimeDerivative
    variable = Temperature
  []
[]

[GrayDiffuseRadiation]
  [vacuum_radiation]
    boundary = 'stator_inner_surface_sideset rotor_outer_surface_sideset'
    emissivity = '0.03 0.03'
    n_patches = '2 2'
    temperature = Temperature
    ray_tracing_face_type = 'GAUSS'
    # fixed_temperature_boundary = 'stator_inner_surface_sideset'
    # fixed_boundary_temperatures = 300
  []
[]

[Materials]
  [thermal_rotor]
    type = HeatConductionMaterial
    thermal_conductivity = 45.0
    specific_heat = 0.5
    block = 'rotor_volume'
  []
  [thermal_stator]
    type = HeatConductionMaterial
    thermal_conductivity = 25.0
    specific_heat = 1
    block = 'stator_volume'
  []
  [density]
    type = GenericConstantMaterial
    prop_names = 'density'
    prop_values = 8000.0
  []
[]

# [BCs]
#  [temperature_stator]
#    type = DirichletBC
#    variable = Temperature
#    value = 300
#    boundary = 'stator_inner_surface'
#  []
# []

[Executioner]
  type = Transient
  end_time = 5
  petsc_options_iname = '-pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter'
  petsc_options_value = 'hypre     boomeramg      5'
  dt = 1
[]

[Outputs]
  exodus = true
[]

```

**Reply by mrplants** [reply]
Beautiful.  Thank you!

**Comment by mrplants** [comment]
Where can I learn more about the various options here?  For example, I've looked in the documentation and "patch" isn't described in detail.  Are there good references where I can learn more about using ray tracing to calculate gray diffuse radiation?

**Reply by snschune** [reply]
We've meant to improve the documentation this year but haven't gotten there yet. Feedback on what is not clear would be helpful.
What we have:
The action is described here:
https://mooseframework.inl.gov/syntax/GrayDiffuseRadiation/index.html
Some info about how view factors are computed:
https://mooseframework.inl.gov/source/userobjects/RayTracingViewFactor.html
https://mooseframework.inl.gov/source/userobjects/ViewFactorRayStudy.html

Patches are simply regions that get an entry in the view factor matrix. Patches are constructed
by subdividing sidesets into smaller sidesets. It's a tradeoff between having a view factor between
each face pair and a view factor between each sideset pair.

**Reply by mrplants** [reply]
How do I get an idea for how many patches I need? Seems like 2 or 20 might be not enough.  I tried scaling it up and it would not go any higher than 1400x1400.  There was no major difference in the results at that scale, though.

**Reply by snschune** [reply]
I would handle it like mesh refinement. Do very few, then double it, and then some more and see if the answer stops changing.
Caveat: the system will work for a moderate number of patches but performance will deteriorate at some point. It's simply the way we deal with patches and how the linear equation for the radiation is solved (essentially in serial). We are going to work on that this year, but it's currently still limited.

**Comment by mrplants** [comment]
By the way, huge thanks to all the folks working on this project.  The Moose Framework seems incredibly powerful from what I can tell.  Not a surprise coming from INL, though.  I used to work with the lab when I was a gov't employee years ago and have been nothing but impressed with the kind of inspiring innovative projects you're all working on.  Thank you so much!  I love Moose!
